### ISSUE TITLE
Report on constructs and operators used

### ISSUE THREAD

**Issue by psychemedia** [dsa-ou/allowed#3]
You could presumably re-use this code to create a simple reporting tool that displays the constructs used in each unit, as well as maintaining "first used in" and "introduced previously" reports.

This would be useful for:

- writing block introductions and summaries;
- summarising pre-requisites (which is to say, previously introduced constructs that are required in the current unit);
- supporting search (e.g. if you list the constructs/operators as metadata in the py/notebook file using `OPERATORS` symbols and providing search over that eg using `jq`)

**Comment by mwermelinger** [comment]
Good ideas, thanks. The various use cases could be distinguished via command line options. The output format would have to change for each case. The name of the script would probably need to change too, or write different scripts with the same core library and configuration file.

I don't understand your 3rd point. Why add metadata (which would easily get out of sync with the notebook content) instead of using the script as the search engine? For example, with a `-s` option for searching we could type `allowed -s 'for-else' path/to/folder` to search for all files that use an else branch in a for-loop.

**Comment by psychemedia** [comment]
Agreed re: the synch issue, although in a notebook you could have a save hook that parses code and updates metadata.

if you have to parse whilst searching, that is expensive, as you need to re-do it every time you search. Unless you have a search engine indexing notebooks (where gain you have the currency issue).

I did start to explore a search engine for notebooks a long time ago - https://github.com/ouseful-testing/nbsearch – but it attracted zero interest. I need to revisit that, and maybe look at other candidate search apps that have appeared in the intervening period.

**Comment by psychemedia** [comment]
Re: pre-requisites — example / use case for how the pre-req might be used in a particular context?
